crates/cxx-qt-lib/src/core/quuid.rs

```diff
@@ -0,0 +1,386 @@
+        ffi::quuid_new_v3(namespace, data)
+    }
+
+    /// On any platform other than Windows, this function returns a new UUID with variant
+    /// `QUuidVariant::DCE` and version `QUuidVersion::Random`. On Windows, a GUID is generated using
+    /// the Windows API and will be of the type that the API decides to create.
+    pub fn new_v4() -> Self {
+        ffi::quuid_new_v4()
+    }
+
+    /// This function returns a new UUID with variant `QUuidVariant::DCE` and version
+    /// `QUuidVersion::Sha1`. `namespace` is the namespace and `data` is the basic data as described
+    /// by RFC 4122.
+    pub fn new_v5(namespace: &Self, data: &[u8]) -> Self {
+        ffi::quuid_new_v5(namespace, data)
+    }
+
+    /// Creates a QUuid object from the string text, which must be formatted as five hex fields
+    /// separated by '-', e.g., "{xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx}" where each 'x' is a hex
+    /// digit. The curly braces shown here are optional, but it is normal to include them.
+    ///
+    /// If the conversion fails, a null UUID is returned.
+    pub fn from_string<'a, S>(s: S) -> Self
+    where
+        S: Into<crate::QAnyStringView<'a>>,
+    {
+        ffi::quuid_from_string(s.into())
+    }
+
+    /// Creates a UUID with the value specified by the parameters.
+    pub const fn from_fields(data1: u32, data2: u16, data3: u16, data4: [u8; 8]) -> Self {
+        Self {
+            data1,
+            data2,
+            data3,
+            data4,
+        }
+    }
+
+    pub const fn as_fields(&self) -> (u32, u16, u16, &[u8; 8]) {
+        (self.data1, self.data2, self.data3, &self.data4)
+    }
+
+    /// Creates a UUID from its representation as a byte array in big endian.
+    pub const fn from_bytes(bytes: [u8; 16]) -> Self {
+        #[cfg(target_endian = "big")]
+        let [b0, b1, b2, b3, b4, b5, b6, b7, b8, b9, ba, bb, bc, bd, be, bf] = bytes;
+        #[cfg(target_endian = "little")]
+        let [b3, b2, b1, b0, b5, b4, b7, b6, b8, b9, ba, bb, bc, bd, be, bf] = bytes;
+        Self {
+            data1: u32::from_ne_bytes([b0, b1, b2, b3]),
+            data2: u16::from_ne_bytes([b4, b5]),
+            data3: u16::from_ne_bytes([b6, b7]),
+            data4: [b8, b9, ba, bb, bc, bd, be, bf],
+        }
+    }
+
+    /// Returns the memory representation of this UUID as a byte array in big-endian byte order.
+    pub const fn to_bytes(&self) -> [u8; 16] {
+        let [b0, b1, b2, b3] = self.data1.to_ne_bytes();
+        let [b4, b5] = self.data2.to_ne_bytes();
+        let [b6, b7] = self.data3.to_ne_bytes();
+        let [b8, b9, ba, bb, bc, bd, be, bf] = self.data4;
+        #[cfg(target_endian = "big")]
+        {
+            [
+                b0, b1, b2, b3, b4, b5, b6, b7, b8, b9, ba, bb, bc, bd, be, bf,
+            ]
+        }
+        #[cfg(target_endian = "little")]
+        {
+            [
+                b3, b2, b1, b0, b5, b4, b7, b6, b8, b9, ba, bb, bc, bd, be, bf,
+            ]
+        }
+    }
```

Comment:
Alternatively, using `std::mem::transmute`:
```rs
    const fn to_big_endian(self) -> Self {
        #[cfg(target_endian = "big")]
        {
            self
        }
        #[cfg(target_endian = "little")]
        {
            Self {
                data1: self.data1.swap_bytes(),
                data2: self.data2.swap_bytes(),
                data3: self.data3.swap_bytes(),
                data4: self.data4,
            }
        }
    }

    /// Creates a UUID from its representation as a byte array in big endian.
    pub const fn from_bytes(bytes: [u8; 16]) -> Self {
        unsafe { mem::transmute::<[u8; 16], Self>(bytes) }.to_big_endian()
    }

    /// Returns the memory representation of this UUID as a byte array in big-endian byte order.
    pub const fn to_bytes(self) -> [u8; 16] {
        unsafe { mem::transmute::<Self, [u8; 16]>(self.to_big_endian()) }
    }
```
It's simpler and more efficient, but at the cost of using `unsafe`, which is why I figured you'd prefer the PR's approach instead.

Reply:
Personally I like the transmute option more. I don't really like the hand-rolled option as that just seems easy to mess up. I wonder if to_big_endian needs/should be public API though?

Reply:
I would be nervous about making `to_big_endian` part of the public API, because it's only a utility for `from_bytes` and `to_bytes`. If we decided to change our implementation of those functions, I personally think we should be able to get rid of `to_big_endian` without it being a breaking change.